api/external_auth.proto

```diff
@@ -0,0 +1,152 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
+
+package envoy.api.v2;
+
+
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "google/rpc/status.proto";
+
+// Authorization service.
+// Primarily responds with OK or NOT OK.
+service Authorization {
+   rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+    // The name of the tenant for multi-tenant support.
+    string tenant_name = 1;
+
+    // The request attributes.
+    AttributeContext attributes = 2;
+}
+
+message CheckResponse {
+    // Status OK allows the request. Any other response indicates the request should be denied.
+    google.rpc.Status status = 1;
+}
+
+// An attribute is a piece of metadata that describes an activity on a network.
+// For example, the size of an HTTP request, or the status code of an HTTP response.
+//
+// Each attribute has a type and a name, which is logically defined as a proto message field.
+// The fields follow the Istio attribute vocabulary.
+// https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html
+// The field type becomes the attribute type, and the field path becomes the attribute name.
+// For example, the attribute `source.ip` maps to field `AttributeContext.source.ip`.
+message AttributeContext {
+  // This message defines attributes for a node that handles a network request.
+  // The node can be either a service or an application that sends, forwards,
+  // or receives the request. Service peers should fill in the `service`,
+  // `principal`, and `labels` as appropriate.
+  message Peer {
+    // The IP address of the peer.
+    string ip = 1;
+
+    // The network port of the peer.
+    int64 port = 2;
+
+    // The canonical service name of the peer.
+    string service = 3;
+
+    // The labels associated with the peer.
+    // These could be pod labels for Kubernetes or tags for VMs.
+    // The source of the labels could be an X.509 certificate or other configuration.
+    map<string, string> labels = 4;
+
+    // The validated identity of this peer.
+    // For example, the identity associated with the workload such as a service account.
+    // If an X.509 certificate is used to assert the identity this field should be sourced from
+    // `Subject` and `Subject Alternative Names`.
+    repeated string principals = 5;
+
+    // Other claims about the peer that are not captured in labels and principals.
+    google.protobuf.Struct claims = 6;
+
+    // The CLDR country/region code associated with the above IP address.
+    // If the IP address is private, the `region_code` should reflect the
+    // physical location where this peer is running.
```

Comment:
where this peer is running, normally as part of configuration of the peer.

The `region_code` is used for jurisdiction policy, e.g. certain features or certain data cannot be used in certain countries or regions.

Reply:
This is another feature that doesn't exist today in Envoy (IP geolocation). I think we want this to be added in the near future (at least Google does). I think this field though would be useful for other purposes (and again, I would like more convergence with logging here), e.g. in a cloud environment it could be used to indicate the originating region/zone.

Reply:
This is not just IP geolocation. When an Envoy server starts, it should know which country it is in (passed via command line flag). IP geolocation is only useful for public ip address.

This feature is required when jurisdiction policy kicks in. We can defer it to a later PR.

Reply:
Let's drop this from the current PR in the interest of keeping scope tight.

Reply:
Fair enough.

Reply:
dropped.